plans/backup.pp

```diff
@@ -68,7 +68,7 @@
     out::message("# Backing up database ${database_names[$index]}")
       # If the primary postgresql host is set then pe-puppetdb needs to be remotely backed up to primary.
       if $database_names[$index] == 'pe-puppetdb' and $cluster['primary_postgresql_host'] {
-        run_command("sudo -u pe-puppetdb /opt/puppetlabs/server/bin/pg_dump \"sslmode=verify-ca host=${cluster['primary_postgresql_host']} sslcert=/etc/puppetlabs/puppetdb/ssl/${primary_host}.cert.pem sslkey=/etc/puppetlabs/puppetdb/ssl/${primary_host}.private_key.pem sslrootcert=/etc/puppetlabs/puppet/ssl/certs/ca.pem dbname=pe-puppetdb\" -f /tmp/puppetdb_$(date +%F_%T).bin" , $primary_host) # lint:ignore:140chars
+        run_command("sudo -u pe-puppetdb /opt/puppetlabs/server/bin/pg_dump \"sslmode=verify-ca host=${cluster['primary_postgresql_host']} sslcert=/etc/puppetlabs/puppetdb/ssl/${primary_host}.cert.pem sslkey=/etc/puppetlabs/puppetdb/ssl/${primary_host}.private_key.pem sslrootcert=/etc/puppetlabs/puppet/ssl/certs/ca.pem dbname=pe-puppetdb\" -f ${backup_directory}/puppetdb_$(date +%F_%T).bin" , $primary_host) # lint:ignore:140chars
```

Comment:
Alternatively... why are we running `pg_dump` as `pe-puppetdb`? Shouldn't we just run this command as the `pe-postgres` user?

Reply:
Yup but as far as i coudld see in the config filed and test

Out of the box config has pe-postgres local access and pe-puppetb remote access

Reply:
Ah, gotcha. Yeah, that makes sense.

I don't like the idea of making pe-puppetdb the directory owner because those seem like objectively incorrect permissions. New tack. What if we either:

1. Run `pg_dump` as root
2. Use a tmpdir and have root move/chown the files after the unprivileged account runs `pg_dump`

For option 1, the main reason to avoid it would be a potential attack vector should a vulnerability in `pg_dump` ever be discovered. Option 2 seems fine security wise, just mildly more work.

The end result though really should be that the directory is owned as root and inaccessible by ALL other accounts. So, we should probably do option 2.

Reply:
I agree I was going quickest fix there but it feels wrong permissions. The problem with option 1 we can't directly run it as root, we end up having to do it as another user, since root has no rights to connect to the database since its not in pg_hba.conf so we end up having to run the commands as a user who does, even if were doing it from root via sudo anyway :-(

So I'll take option 2 👍